### DISCUSSION TITLE
Design and Build CLI

### DISCUSSION THREAD

**Discussion by lesagi** [mini-services/miniurl#153] [General]
This discussion is for deciding on the following:
1. What functions should the CLI serve?
2. Suggesting its design based on the previous point
3. Choosing the right framework to use for this task

As I see it, we'd eventually like to support these features:
1. as long all we have is the miniurl:
    1. setting the configuration for the project (whether through CLI, or UI interface)
2. after generalizing the service,
    1. bootstrapping a mini services infrastructure (with a default template/custom configuration of the infra (such as InMemory DB, Relational, etc..)

I think we should list all features that we'd want to support in the near future, and then moving forward to design.
I checked the following Frameworks for building the CLI, let me know if there another one recommended:
1. Commander
2. Inquirer/enquirer
3. Listr
4. progress/cli-progress
5. oclif 
6. gluegun

Above all, I liked 'Inquirer' the most, but if anyone of you wants to recommend a different one, let me know

**Comment by SnirShechter** [comment]
In my opinion, these changes should be made
### Features
* Generate config - we need to find out the format we want the config to be. If it's a JS/TS file, we need to implement a way to use it in MiniUrl (currently it works via environment variables only).
* Interfaces - I'd say we want three interfaces
    * CLI commands e.g `mini miniurl generate-config --yes` (Commander style)
    * CLI interactive questions (Inquirer style)
    * UI - an app that will open in a browser and have an interactive and descriptive way of configuring the project (Vue CLI style). ***This should be delayed for later.***

### Libraries
* After stating the above, I believe Commander is the best option for the CLI commands, Inquirer is the best option for CLI interactive questions and Listr/progress should be used for progress/loaders.

### Action items *(after deciding on the design & libraries)*
* Create an empty TS project with the basic boilerplate:
    * Typescript (note that the CLI will be an executable so it should be compiled to JS unlike MiniUrl)
    * Linting - ESLint, Prettier, Lint-staged, yorkie
    * Testing - Jest, codecov
    * Logging - Pino.js
* Design the folder structure and functionalities:
    * Where will all the commands be? a file for each command? what about nested commands? how can we namespace commands for the different future microservices?
    * What functionality should each command implement? How do we avoid writing completely different CLIs for each interface?
* Create a ticket regarding the config implementation in MiniUrl, design it, and find someone to work on it.
